moveit_planners/ompl/ompl_interface/src/detail/ompl_constraints.cpp

```diff
@@ -385,4 +385,101 @@ std::shared_ptr<BaseConstraint> createOMPLConstraint(const moveit::core::RobotMo
     return nullptr;
   }
 }
-}  // namespace ompl_interface
\ No newline at end of file
+
+/******************************************
+ * Toolpath Constraint
+ * ****************************************/
+ToolPathConstraint::ToolPathConstraint(const moveit::core::RobotModelConstPtr& robot_model, const std::string& group,
+                                       unsigned int num_dofs, unsigned int steps)
+  : BaseConstraint(robot_model, group, num_dofs), num_steps_(steps)
+{
+  NearestNeighbors<geometry_msgs::msg::Pose>::DistanceFunction func =
+      std::bind(&poseDistance, std::placeholders::_1, std::placeholders::_2);
+
+  pose_nn_.setDistanceFunction(func);
+}
+
+void ToolPathConstraint::setPath(const moveit_msgs::msg::CartesianTrajectory& msg)
+{
+  // TODO: Should this be done in the constructor? What happens when we want to accept another msg?
+
+  // Clear previous path
+  pose_nn_.clear();
+
+  // Interpolate along waypoints and insert into pose_nn_;
+  const auto dt = 1.0 / static_cast<double>(num_steps_);
+  for (unsigned int i = 0; i < msg.points.size() - 1; i++)
+  {
+    // Insert waypoints
+    pose_nn_.add(msg.points.at(i).point.pose);
+
+    // Insert interpolated poses
+    auto step = dt;
+    for (unsigned int j = 0; j < num_steps_ - 1; j++)
+    {
+      const auto pose = poseInterpolator(msg.points.at(i).point.pose, msg.points.at(i + 1).point.pose, step);
+      pose_nn_.add(pose);
+      step += dt;
+    }
+  }
+
+  // Insert last waypoint
+  pose_nn_.add(msg.points.at(msg.points.size() - 1).point.pose);
+}
+
+
+void ToolPathConstraint::function(const Eigen::Ref<const Eigen::VectorXd>& joint_values, Eigen::Ref<Eigen::VectorXd> out) const
+{
+  // TODO: Add path constraint function
+  // Ideal: F(q_actual) - F(q_neared_pose) = 0
+
+}
+
+
+
+double poseDistance(const geometry_msgs::msg::Pose& pose1, const geometry_msgs::msg::Pose& pose2)
+{
+  const auto dx = pose1.position.x - pose2.position.x;
+  const auto dy = pose1.position.y - pose2.position.y;
+  const auto dz = pose1.position.z - pose2.position.z;
+
+  // Distance between two quaternions
+  // https://math.stackexchange.com/questions/90081/quaternion-distance
+
+  // dot product between q1 and q2
+  const auto qdotprod = pose1.orientation.w * pose2.orientation.w + pose1.orientation.x * pose2.orientation.x +
+                        pose1.orientation.y * pose2.orientation.y + pose1.orientation.z * pose2.orientation.z;
+
+  // Euclidean + Quaternion Distance
+  return std::sqrt(dx * dx + dy * dy + dz * dz) + (1.0 - qdotprod * qdotprod);
```

Comment:
quaternion distance is not right. See https://github.com/ompl/ompl/blob/master/src/ompl/base/spaces/src/SO3StateSpace.cpp#L254-L262
The `fabs` and `acos` are both important.